### PR TITLE
Support Custom Sentiment Analysis task (LRO) in AnalyzeText Language API 

### LINKED ISSUE #22490: [Cognitive Services - Text Analytics API] t Custom Sentiment Analysis task API Review
New API Review meeting has been requested.

**Service Name**: Cognitive Services Text Analytics API
**Review Created By**: Ranran Wang @jessicawrr
**Review Date**: 3/2/2023 4:00 PM PT
**PR**: 
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: 

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=5712f885-1da7-ed11-aad1-000d3a3418aa)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by jessicawrr**
# Data Plane API - Pull Request

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document:
* Previous Open API Doc:
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

### CODE REVIEW

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kristapratico) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1665]: Will CustomSentimentAnalysis support opinion mining (now or in the future)?

File: dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskSubmit.json
Comment (kristapratico) [inline dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskSubmit.json:30]: Should the api version be 2023-04-15-preview?

```suggestion
        "Operation-Location": "{Endpoint}/language/analyze-text/jobs/{jobId}?api-version=2023-04-15-preview"
```

**Review by heaths — APPROVED** [review]

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (jessicawrr) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1665]: not in the near future. That's why we don't introduce the om parameter here

File: dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskSubmit.json
Comment (jessicawrr) [reply dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskSubmit.json:30]: Good catch!! Will update

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1700]: This object also describes the results for opinion mining and you said opinion mining is not part of this feature. This sharing could cause confusion. Could you please define your own result object that has things that are unique to this task only?

Comment (jessicawrr) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1700]: Sounds good. Will do that!

**Comment by heaths** [comment]
Please rebase your changes on the target release branch and don't merge main. It's impossible to review this:

image

```bash
git fetch origin release-cognitiveservices-Language-2023-04-15-preview
# on your topic branch...
git rebase origin/release-cognitiveservices-Language-2023-04-15-preview
git push --force
```

That will update this PR so please don't close this one and create a new one.

**Comment by heaths** [comment]
Can you rebase this?

image

```bash
git fetch upstream release-cognitiveservices-Language-2023-04-15-preview
# on your topic branch
git rebase FETCH_HEAD
git push --force
```

**Review by markweitzel — APPROVED** [review]
## API Stewardship Board
Per the review on 2-Mar-23, this is good to go for preview. 
Note: Heath will partner with the team to clean up the branch.

**Review by heaths — CHANGES_REQUESTED** [review]
Need to clean up the branch before merging. @Ransqure can you confirm that only the following two commits are needed?

```
commit df892ba48f2c2788d5ce0fd84488f85d572c918f
Author: Ranran Wang <rawan@microsoft.com>
Date:   Sun Feb 5 20:56:59 2023 -0800
    Adding Custom Sentiment task request and response example

commit f61a5f14a0a73cab833f9deba2c17a0dc8d1a629
Author: Ranran Wang <rawan@microsoft.com>
Date:   Sun Feb 5 15:52:57 2023 -0800
    Adding Custom Sentiment LRO task and its response
```

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (heaths) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1700]: @jessicawrr any update on resolving this issue?

**Review by heaths — CHANGES_REQUESTED** [review]
@deyaaeldeen called out a change needed before I can merge.

**Comment by heaths** [comment]
Please also fix the prettier issue. See https://github.com/Azure/azure-rest-api-specs/pull/22466/checks?check_run_id=11982254371 for details.

As for breaking changes ([policy](https://aka.ms/AzBreakingChangesPolicy)),

1. Don't add `type: object` to previously GA'd models. It would've been better to have them initially but, practically, it doesn't make a difference for our code generators. Adding them now is considered a breaking change.
2. Based on [the diff to 2022-05-01](https://portal.azure-devex-tools.com/tools/diff/eyJzb3VyY2VVcmwiOiJodHRwczovL3Jhdy5naXRodWJ1c2VyY29udGVudC5jb20vQXp1cmUvYXp1cmUtcmVzdC1hcGktc3BlY3MvbWFpbi9zcGVjaWZpY2F0aW9uL2NvZ25pdGl2ZXNlcnZpY2VzL2RhdGEtcGxhbmUvTGFuZ3VhZ2Uvc3RhYmxlLzIwMjItMDUtMDEvYW5hbHl6ZXRleHQuanNvbiIsInRhcmdldFVybCI6Imh0dHBzOi8vcmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbS9BenVyZS9henVyZS1yZXN0LWFwaS1zcGVjcy8yYzMwODQxZGYzYWVjZWM2NWEzNjM0ZTMwYWEyMDVlOTNmOTkyZDkxL2Rldi9jb2duaXRpdmVzZXJ2aWNlcy9kYXRhLXBsYW5lL0xhbmd1YWdlL2FuYWx5emV0ZXh0Lmpzb24ifQ==), the newly added required properties on previously GA'd models are all on response models, so allowed by the Breaking Change Policy since this is a new api-version. That said, some models updated `$ref`s to point to similar models in _common.json_. at least from what I can see in the [corresponding _common.json_](https://github.com/Azure/azure-rest-api-specs/blob/2c30841df3aecec65a3634e30aa205e93f992d91/dev/cognitiveservices/data-plane/Language/common.json) these appear to be equivalent. Most of these validation errors may be from between the new and old versions of 2023-04-15-preview, which validation shouldn't really be running on anyway since it's **in-development**. I opened Azure/azure-sdk-tools#5713 to track also adding links or at least a model name to make it more obvious where the validation errors are.

**Comment by deyaaeldeen** [comment]
What is the status for this PR?

**Comment by heaths** [comment]
`npx prettier@2.6.2 -c dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskSubmit.json` shows no issues on the checked out files for this PR. I'll try to run failed checks again; otherwise, need @lmazuel or @weshaggard to use their special powers to merge. This is not a check I can override with a label.

**Comment by jessicawrr** [comment]
Hi everyone. Sorry for the late response. We need to merge this PR in. I believe I address all the comments. @heaths is there anything else need to address before merging it?

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1700]: @jessicawrr could you please address this comment?

**Comment by heaths** [comment]
@jessicawrr you'll need to resolve the conflict in analyzetext.json, but @lmazuel or @weshaggard will have to merge. I can't because the PrettierCheck is required and I can't determine why. There's no output in the logs, and when I try it locally against this PR I also see no output. No idea what's wrong.

**Comment by weshaggard** [comment]
@heaths what about the breaking change?

**Comment by heaths** [comment]
> what about the breaking change?

@weshaggard they are all in response models where `required` doesn't make much difference in most languages' code generators, but where for newly-added properties anyway.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (jessicawrr) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json]: Oh Sorry. I missed this. I've updated the PR and defined the custom sa specific result

**Comment by jessicawrr** [comment]
@heaths and @deyaaeldeen I addressed the comments you point out. Also. I've resolved the conflict in analyzetext.json. I believe we are good to merge it :)

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (weshaggard) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:318]:
```suggestion
        "CustomSentimentAnalysis",
```

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (weshaggard) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:359]:
```suggestion
        "CustomSentimentAnalysisLROResults",
```

**Comment by heaths** [comment]
There an unresolvable reference that must be fixed before we can merge or it'll break everything for this service directory.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1429]: nitpick: Would be nice if we align the custom names with the prebuilt names.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json]: Is this an orphan definition? I don't see it referenced

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (mssfang) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:172]: NIT: "Custom" is repeating.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (mssfang) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1788]: should we align with other same ref. 
See https://github.com/Azure/azure-rest-api-specs/blob/c34abf84ee5f0e20f599036184e9817da49d1945/dev/cognitiveservices/data-plane/Language/analyzetext.json#L4094


 - $ref: common.json#/definitions/CustomResult

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (mssfang) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1749]: See https://github.com/Azure/azure-rest-api-specs/blob/c34abf84ee5f0e20f599036184e9817da49d1945/dev/cognitiveservices/data-plane/Language/analyzetext.json#L586

**Comment by konrad-jamrozik** [comment]
> `npx prettier@2.6.2 -c dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskSubmit.json` shows no issues on the checked out files for this PR. I'll try to run failed checks again; otherwise, need @lmazuel or @weshaggard to use their special powers to merge. This is not a check I can override with a label.

@heaths looks like the problem is with a different file: [`*Request.json` instead of `*Submit.json`](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2684744&view=logs&j=781fd7cb-cbf2-5d4a-a595-f7c2fa08ccdd&t=32535914-32d8-5ecd-c7d0-aa49d9c79250&l=15):

``` powershell
npx prettier -c "dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskStatusRequest.json"
Checking formatting...
[warn] dev\cognitiveservices\data-plane\Language\examples\text\SuccessfulCustomSentimentAnalysisTaskStatusRequest.json
[warn] Code style issues found in the above file. Forgot to run Prettier?
```

@jessicawrr you can fix the issue by running following command in the local clone of your fork on branch `user/rawan/customsa-runtime-api`, and pushing the resulting changes:

``` powershell
npx prettier -w "dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskStatusRequest.json"
```

**Comment by jessicawrr** [comment]
I see. Let me try

**Comment by heaths** [comment]
@konrad-jamrozik @weshaggard the netcoreapp3.1 issue is still a blocker. It can't run the breaking change validation until that's fixed or this PR is rebased, which it may need to be anyway. Let me know, and I can help @jessicawrr if needed.

@jessicawrr  you also need to resolve the conflicts since another version of _analyzetext.json_ went in already.

image

**Comment by weshaggard** [comment]
@heaths that should be fixed when a new pipeline is triggered after the merge conflicts are resolved.

**Comment by jessicawrr** [comment]
@heaths I believe I resolved everything blocking the merge (resolving the conflict and fix the PrettierCheck). Please take a look.

**Review by heaths — APPROVED** [review]

**Pull request merged**
